### PR TITLE
Henryh/per modality with pooling

### PR DESCRIPTION
**Pull request by Hgherzog**
- add registers to both the cross pooled model and the normal architecture

### CODE REVIEW

File: .github/workflows/tests.yaml [hunk 1]
```diff
@@ -4,13 +4,14 @@ on:
   push:
     branches:
       - main
+      - henryh/per-modality-with-pooling
```
Comment (yawenzzzz) [inline .github/workflows/tests.yaml]: We can remove this once this PR gets merged.

File: helios/train/loss.py [hunk 2]
```diff
@@ -288,6 +294,7 @@ def __init__(self, tau: float = 0.1, pred2unit: bool = False, weight: float = 1.
         self.tau = tau
         self.pred2unit = pred2unit
         self.weight = weight
+        self.modality_weights = modality_weights
```
Comment (yawenzzzz) [inline helios/train/loss.py:298]: Suggestion: add this into the Args

File: helios/train/masking.py [hunk 3]
```diff
@@ -180,6 +180,39 @@ def from_dict(cls, dict: dict[str, Any]) -> "MaskedHeliosSample":
...
+            logger.info(f"Per modality present masks shape: {present_mask.shape}")
+            per_modality_present_masks.append(present_mask)
+
+        at_least_one_modality_present_timestep_mask = (
+            torch.stack(per_modality_present_masks, dim=1).any(dim=1).squeeze(1)
+        )
+        logger.info(
+            f"At least one modality present timestep mask shape: {at_least_one_modality_present_timestep_mask.shape}"
+        )
+        return at_least_one_modality_present_timestep_mask
```
Comment (yawenzzzz) [inline helios/train/masking.py]: It's fine to add this in `MaskedHeliosSample`, but seems `timesteps_with_at_least_one_modality` is only used in `HeliosSample`

File: helios/nn/flexihelios.py [hunk 4]
```diff
@@ -995,7 +1028,10 @@ def stack_spatial_modalities_and_masks(
         mask_list = []
         data_list = []
         for modality in modalities_to_process:
-            if Modality.get(modality).is_spatial:
+            # Hack for now only encoding spatial temporal modalities
+            modality_spec = Modality.get(modality)
+            if modality_spec.is_spatial and modality_spec.is_multitemporal:
+                logger.info(f"modality: {modality} is spatial")
```
Comment (yawenzzzz) [inline helios/nn/flexihelios.py]: spatial -> spatial and temporal

File: helios/nn/flexihelios.py [hunk 5]
```diff
@@ -185,6 +185,74 @@ def flatten_tokens_and_masks(
...
+        if len(spatial_average) == 0:
+            raise ValueError("Missing unmasked spatial modalities for spatial pooling.")
+        spatial_average_t = torch.stack(spatial_average, dim=-1)
+        if pooling_type == PoolingType.MEAN:
+            return spatial_average_t.mean(dim=-1)
+        else:
+            return spatial_average_t.max(dim=-1).values
+
+    def pool_instance_wise(self, pooling_type: PoolingType) -> Tensor:
+        """Pool al the tokens in the instance."""
```
Comment (yawenzzzz) [inline helios/nn/flexihelios.py]: Fix typo here.

File: helios/nn/flexihelios.py [hunk 6]
```diff
@@ -202,74 +270,13 @@ def pool_unmasked_tokens(
...
-            if len(spatial_average) == 0:
-                raise ValueError(
-                    "Missing unmasked spatial modalities for spatial pooling."
-                )
-            spatial_average_t = torch.stack(spatial_average, dim=-1)
-            if pooling_type == PoolingType.MEAN:
-                return spatial_average_t.mean(dim=-1)
-            else:
-                return spatial_average_t.max(dim=-1).values
+            return self.pool_spatially(pooling_type)
```
Comment (yawenzzzz) [inline helios/nn/flexihelios.py:279]: This is nice!

File: helios/nn/flexihelios.py [hunk 7]
```diff
@@ -297,17 +304,25 @@ def __init__(
...
+        self, x: TokensAndMasks | torch.Tensor
+    ) -> torch.Tensor:
+        """Apply the aggregate operation to the input."""
+        if isinstance(x, TokensAndMasks):
             pooled_for_contrastive = x.pool_unmasked_tokens(
                 PoolingType.MEAN, spatial_pooling=False
             )
-            return self.projection(pooled_for_contrastive)
+        elif isinstance(x, torch.Tensor):
+            pooled_for_contrastive = reduce(x, "b ... d -> b  d", "mean")
```
Comment (yawenzzzz) [inline helios/nn/flexihelios.py:316]: Do we want to have `max` reduce as well?

File: helios/train/masking.py (hunk 3 above)
Comment (Hgherzog) [reply helios/train/masking.py]: this can be deleted

File: helios/nn/flexihelios.py (hunk 7 above)
Comment (Hgherzog) [reply helios/nn/flexihelios.py:316]: We have only supported mean in the past. Not going to add until we try it.
